suitesparse_graphblas/__init__.py

```diff
@@ -37,8 +41,176 @@ def initialize(*, blocking=False, memory_manager="numpy"):
+    try:
+        return getattr(lib, name)
+    except AttributeError:
+        ext_name = f"GxB_{name[4:]}"
+        try:
+            return getattr(lib, ext_name)
+        except AttributeError:
+            pass
+        raise
+
+
+bool_types = (lib.GrB_BOOL,)
+
+signed_integer_types = (
+    lib.GrB_INT8,
+    lib.GrB_INT16,
+    lib.GrB_INT32,
+    lib.GrB_INT64,
+)
+
+unsigned_integer_types = (
+    lib.GrB_UINT8,
+    lib.GrB_UINT16,
+    lib.GrB_UINT32,
+    lib.GrB_UINT64,
+)
+
+integer_types = signed_integer_types + unsigned_integer_types
+
+real_types = (
+    lib.GrB_FP32,
+    lib.GrB_FP64,
+)
+
+if supports_complex():
+    complex_types = (
+        lib.GxB_FC32,
+        lib.GxB_FC64,
+    )
+else:
+    complex_types = ()
+
+
+grb_types = bool_types + integer_types + real_types + complex_types
+
+
+class GraphBLASException(Exception):
+    pass
+
+
+class NoValue(GraphBLASException):
+    pass
+
+
+class UninitializedObject(GraphBLASException):
+    pass
+
+
+class InvalidObject(GraphBLASException):
+    pass
+
+
+class NullPointer(GraphBLASException):
+    pass
+
+
+class InvalidValue(GraphBLASException):
+    pass
+
+
+class InvalidIndex(GraphBLASException):
+    pass
+
+
+class DomainMismatch(GraphBLASException):
+    pass
+
+
+class DimensionMismatch(GraphBLASException):
+    pass
+
+
+class OutputNotEmpty(GraphBLASException):
+    pass
+
+
+class OutOfMemory(GraphBLASException):
+    pass
+
+
+class InsufficientSpace(GraphBLASException):
+    pass
+
+
+class IndexOutOfBound(GraphBLASException):
+    pass
+
+
+class Panic(GraphBLASException):
+    pass
+
+
+_error_code_lookup = {
+    # Warning
+    lib.GrB_NO_VALUE: NoValue,
+    # API Errors
+    lib.GrB_UNINITIALIZED_OBJECT: UninitializedObject,
+    lib.GrB_INVALID_OBJECT: InvalidObject,
+    lib.GrB_NULL_POINTER: NullPointer,
+    lib.GrB_INVALID_VALUE: InvalidValue,
+    lib.GrB_INVALID_INDEX: InvalidIndex,
+    lib.GrB_DOMAIN_MISMATCH: DomainMismatch,
+    lib.GrB_DIMENSION_MISMATCH: DimensionMismatch,
+    lib.GrB_OUTPUT_NOT_EMPTY: OutputNotEmpty,
+    # Execution Errors
+    lib.GrB_OUT_OF_MEMORY: OutOfMemory,
+    lib.GrB_INSUFFICIENT_SPACE: InsufficientSpace,
+    lib.GrB_INDEX_OUT_OF_BOUNDS: IndexOutOfBound,
+    lib.GrB_PANIC: Panic,
+}
+GrB_SUCCESS = lib.GrB_SUCCESS
+GrB_NO_VALUE = lib.GrB_NO_VALUE
+
+
+_error_func_lookup = {
+    "struct GB_Type_opaque *": lib.GrB_Type_error,
+    "struct GB_UnaryOp_opaque *": lib.GrB_UnaryOp_error,
+    "struct GB_BinaryOp_opaque *": lib.GrB_BinaryOp_error,
+    "struct GB_SelectOp_opaque *": lib.GxB_SelectOp_error,
+    "struct GB_Monoid_opaque *": lib.GrB_Monoid_error,
+    "struct GB_Semiring_opaque *": lib.GrB_Semiring_error,
+    "struct GB_Scalar_opaque *": lib.GxB_Scalar_error,
+    "struct GB_Matrix_opaque *": lib.GrB_Matrix_error,
+    "struct GB_Vector_opaque *": lib.GrB_Vector_error,
+    "struct GB_Descriptor_opaque *": lib.GrB_Descriptor_error,
+}
```

Comment:
Should we also include strings such as `"struct GB_Vector_opaque"`?  Or, if I happened to pass an object of this type in, would I be doing things wrong?

Reply:
Yeah, I think we should have the non-pointer versions as keys in this dict.